sdk/core/Azure.Core/src/Shared/ConnectionString.cs

```diff
@@ -0,0 +1,248 @@
+                else
+                {
+                    stringBuilder.Append(_segmentSeparator);
+                }
+
+                stringBuilder.Append(segment.Key);
+                if (segment.Value != null)
+                {
+                    stringBuilder.Append(_keyValueSeparator).Append(segment.Value);
+                }
+            }
+
+            return stringBuilder.ToString();
+        }
+
+
+        private static Dictionary<string, string> ParseSegments(in string connectionString, in string segmentSeparator, in string keyValueSeparator)
+        {
+            var segments = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            var segmentStart = -1;
+            var segmentEnd = 0;
+
+            while (TryGetNextSegment(connectionString, segmentSeparator, ref segmentStart, ref segmentEnd))
+            {
+                var kvSeparatorIndex = connectionString.IndexOf(keyValueSeparator, segmentStart, segmentEnd - segmentStart, StringComparison.Ordinal);
+                int keyStart, keyLength;
+                if (kvSeparatorIndex != -1)
+                {
+                    keyStart = GetStart(connectionString, segmentStart);
+                    keyLength = GetLength(connectionString, keyStart, kvSeparatorIndex);
+                }
+                else
+                {
+                    keyStart = GetStart(connectionString, segmentStart);
+                    keyLength = GetLength(connectionString, keyStart, segmentEnd);
+                }
+
+                var key = connectionString.Substring(keyStart, keyLength);
+                if (segments.ContainsKey(key))
+                {
+                    throw new InvalidOperationException($"Duplicated key '{key}'");
+                }
+
+                if (kvSeparatorIndex != -1)
+                {
+                    var valueStart = GetStart(connectionString, kvSeparatorIndex + keyValueSeparator.Length);
+                    var valueLength = GetLength(connectionString, valueStart, segmentEnd);
+                    segments.Add(key, connectionString.Substring(valueStart, valueLength));
+                }
+                else
+                {
+                    segments.Add(key, null);
+                }
+            }
+
+            return segments;
+
+            static int GetStart(in string str, int start)
+            {
+                while (start < str.Length && char.IsWhiteSpace(str[start]))
+                {
+                    start++;
+                }
+
+                return start;
+            }
+
+            static int GetLength(in string str, in int start, int end)
+            {
+                while (end > start && char.IsWhiteSpace(str[end - 1]))
+                {
+                    end--;
+                }
+
+                return end - start;
+            }
+        }
+
+        private static bool TryGetNextSegment(in string str, in string separator, ref int start, ref int end)
```

Comment:
You shouldn't need in for the reference types parameters.

Reply:
yes, but that ensures that I don't change parameters inside method body.